pkg/controller/kubedirectorcluster/cluster_restore.go

```diff
@@ -0,0 +1,411 @@
+			if annVal == "false" {
+				shared.LogInfo(
+					reqLogger,
+					cr,
+					shared.EventReasonCluster,
+					"being restored: kdstatusbackup was not ever created",
+				)
+				cr.Status.RestoreProgress.Error =
+					"KD status backups were not enabled at the time this kdcluster " +
+						"was snapshotted; it is not possible to do automatic status " +
+						"recovery and resume reconciliation."
+			}
+		}
+	}
+
+	statusBackup, err := observer.GetStatusBackup(
+		cr.Namespace,
+		cr.Name,
+	)
+
+	if err == nil {
+		if cr.Status.RestoreProgress.AwaitingStatus {
+			cr.Status.RestoreProgress.AwaitingStatus = false
+			shared.LogInfo(
+				reqLogger,
+				cr,
+				shared.EventReasonCluster,
+				"being restored: kdstatusbackup is present",
+			)
+			statusBackup.Spec.StatusBackup.RestoreProgress = cr.Status.RestoreProgress
+			cr.Status = statusBackup.Spec.StatusBackup
+		}
+	} else {
+		cr.Status.RestoreProgress.AwaitingStatus = true
+		shared.LogInfo(
+			reqLogger,
+			cr,
+			shared.EventReasonCluster,
+			"being restored: awaiting kdstatusbackup",
+		)
+	}
+}
+
+// checkResourcesRestored looks to see if resources named in the status,
+// which KD is responsible for directly creating, all exist. Set the
+// restoreProgress.awaitingResources flag accordingly.
+func checkResourcesRestored(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) {
+
+	// We need to check for the cluster (headless) service, the statefulsets,
+	// and the per-member services associated with statefulset pods.
+	// Everything else will be created by the statefulset controller.
+
+	resourcesPresent := clusterServiceExists(reqLogger, cr)
+	resourcesPresent = resourcesPresent && roleResourcesExist(reqLogger, cr)
+
+	if resourcesPresent {
+		if cr.Status.RestoreProgress.AwaitingResources {
+			cr.Status.RestoreProgress.AwaitingResources = false
+			shared.LogInfo(
+				reqLogger,
+				cr,
+				shared.EventReasonCluster,
+				"being restored: component resources are present",
+			)
+		}
+	} else {
+		cr.Status.RestoreProgress.AwaitingResources = true
+		shared.LogInfo(
+			reqLogger,
+			cr,
+			shared.EventReasonCluster,
+			"being restored: awaiting component resources",
+		)
+	}
+}
+
+// clusterServiceExists looks to see if a headless service named in the
+// status exists.
+func clusterServiceExists(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+
+	clusterService := cr.Status.ClusterService
+	if clusterService != "" {
+		_, serviceErr := observer.GetService(
+			cr.Namespace,
+			clusterService,
+		)
+		if serviceErr != nil {
+			shared.LogInfof(
+				reqLogger,
+				cr,
+				shared.EventReasonCluster,
+				"being restored: cluster service %s does not exist",
+				clusterService,
+			)
+			return false
+		}
+	}
+	return true
+}
+
+// roleResourcesExist looks to see if a statefulsets named in the
+// status exist, along with the necessary per-member services.
+func roleResourcesExist(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
```

Comment:
what about _connection_ resources? Like configmap, secret or another kdcluster?  what happens if it was connected to another kdcluster and that is not yet up ?

Reply:
Yeah that's a good question. Initially you'd have the same situation as if the configmap (or kdcluster or whatever) had not been created yet, i.e. it would be omitted from the configmeta sent into the container. Then when the object comes back, our connections hash will change, we'll modify the configmeta to include it, and do a reconnect notify.

If an app doesn't support the reconnect notify and always expects its connections to be there before it gets itself configured, then it could be upset at that. Do you think we need the semantic where reconciliation is blocked until connections come back? A potential kicker with that, is that there might have been a missing connection to begin with (at the time the kdcluster was backed up) and so that one will _never_ "come back" on the restore.

Reply:
Hmmm, currently most of the apps do not implement --reconnect since it was added later. Hence it will be a common scenario which we should worry about. So, yes blocking restore until all connections are not up will be cleaner. Can we check for missing connection at the backup time also and if there is then we store that info somewhere. 
This can be very involved and we can do this in another PR but someone like GL would definitely need this if they want to utilize backup functionality for existing apps like notebook with connections.

Reply:
On similar lines, will sequencing matter for these features - 
1. blockStorageDevice
2. serviceAccountName
3. secret mount

Reply:
In general for things like that we shouldn't be concerned. Those are items that went into constructing the pod template inside the statefulset spec, and our philosophy about backup/restore should be that making the native K8s objects work correctly after restoration is up to the backup/restore solution to handle.

For example if an indicated service account is missing, the statefulset controller will fail to create a pod until the service account exists. If a secret-to-mount is missing, the sset controller can create the pod but it won't get out of ContainerCreating state until the secret exists. In those cases ordering doesn't really matter a lot as the pod container won't finally come up until those things are present.

For the block storage, similar to the other filesystem-mode persistent storage, Velero should recreate the PVs before recreating the statefulset (with appropriate claimRefs pointing back to the PVCs that the statefulset controller will create). I'm not sure what other backup solutions will do but I guess as mentioned above this is the sort of thing I would put "below the line" of KD's concern. I.e. anything that would have to be taken care of to make a "normal" statefulset work, we can depend on for the statefulsets that make up KD roles.